staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -82,57 +82,88 @@ type kmsv2PluginProbe struct {
 	l            *sync.Mutex
 }
 
+type kmsHealthChecker []healthz.HealthChecker
+
+func (k kmsHealthChecker) Name() string {
+	return "kms-providers"
+}
+
+func (k kmsHealthChecker) Check(req *http.Request) error {
+	for i := range k {
+		checker := k[i]
+		if err := checker.Check(req); err != nil {
+			return fmt.Errorf("%s healthz check failed: %w", checker.Name(), err)
+		}
+	}
+
+	return nil
+}
+
 func (h *kmsPluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 	return healthz.NamedCheck(fmt.Sprintf("kms-provider-%d", idx), func(r *http.Request) error {
 		return h.check()
 	})
 }
 
-func (p *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
+func (h *kmsv2PluginProbe) toHealthzCheck(idx int) healthz.HealthChecker {
 	return healthz.NamedCheck(fmt.Sprintf("kms-provider-%d", idx), func(r *http.Request) error {
-		return p.check(r.Context())
+		return h.check(r.Context())
 	})
 }
 
-func LoadEncryptionConfig(filepath string, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, error) {
+// LoadEncryptionConfig parses and validates the encryption config specified by filepath.
+// It may launch multiple go routines whose lifecycle is controlled by stopCh.
+// If reload is true, or KMS v2 plugins are used, the return slice of health checkers will always be of length 1.
+func LoadEncryptionConfig(filepath string, reload bool, stopCh <-chan struct{}) (map[schema.GroupResource]value.Transformer, []healthz.HealthChecker, error) {
 	config, err := loadConfig(filepath)
 	if err != nil {
-		return nil, nil, fmt.Errorf("error while parsing file: %v", err)
+		return nil, nil, fmt.Errorf("error while parsing file: %w", err)
+	}
+
+	transformers, kmsHealthChecks, kmsv2, err := getTransformerOverridesAndKMSPluginHealthzCheckers(config, stopCh)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error while building transformers: %w", err)
 	}
 
-	return getTransformerOverridesAndKMSPluginHealthzCheckers(config, stopCh)
+	if reload || kmsv2 {
```

Comment:
collapsing to a single health check when reloading makes sense

collapsing to a single health check when not reloading and *only* using kmsv2 also makes sense (kmsv2 removes the need to have multiple entries during a key rotation... having multiple kmsv2 entries just to involve different kms backends in encrypting different resources and needing to monitor them distinctly via kube-apiserver healthz checks seems... far-fetched to me)

if you're not reloading and still have kmsv1 providers, I think we should keep the distinct checks. That lets kmsv1 users still distinguish between kmsv1 and kmsv2 until they _complete_ the transition to just kmsv2.